parsers/NTESMO.py

```diff
@@ -0,0 +1,214 @@
+            )
+            index[date.month][date.day] = a["href"]
+    return index
+
+
+def get_historical_daily_data(link: str, session: Session) -> bytes:
+    result = session.get(link)
+    return result.content
+
+
+def extract_production_data(file: bytes) -> pd.DataFrame:
+    return pd.read_excel(
+        file, "Generating Unit Output", skiprows=4, header=0, usecols="A:AA"
+    )
+
+
+def extract_demand_price_data(file: bytes) -> pd.DataFrame:
+    return pd.read_excel(
+        file, "System Demand and Market Price", skiprows=4, header=0, usecols="A:C"
+    )
+
+
+def get_data(
+    session: Session,
+    target_datetime: Optional[datetime],
+    extraction_func: Callable[[bytes], pd.DataFrame],
+    logger: Logger,
+) -> pd.DataFrame:
+    if target_datetime is None:
+        target_datetime = datetime.utcnow()
+    index = construct_year_index(target_datetime.year, session)
+    try:
+        data_file = get_historical_daily_data(
+            index[target_datetime.month][target_datetime.day], session
+        )
+    except KeyError:
+        raise Exception(
+            f"Cannot find file on the index page for date {target_datetime}"
+        )
+    return extraction_func(data_file)
+
+
+def parse_consumption(
+    raw_consumption: pd.DataFrame,
+    target_datetime: datetime,
+    logger: Logger,
+    price: bool = False,
+) -> List[dict]:
+    data_points = []
+    for _, consumption in raw_consumption.iterrows():
+        # Market day starts at 4:30 and reports up until 4:00 the next day.
+        # Therefore timestamps between 0:00 and 4:30 excluded need to have an extra day.
+        raw_timestamp = consumption[0]
+        timestamp = datetime.combine(date=target_datetime.date(), time=raw_timestamp)
+        if raw_timestamp < time(hour=4, minute=30):
+            timestamp = timestamp + timedelta(days=1)
+
+        data_point = {
+            "zoneKey": "AU-NT",
+            "datetime": timestamp,
+            "source": "ntesmo.com.au",
+        }
+        if price:
+            data_point["price"] = consumption["Market Price"]
+            data_point["currency"] = "AUD"
+        else:
+            data_point["consumption"] = consumption["Demand"]
+        data_points.append(data_point)
+    return data_points
+
+
+def parse_production_mix(
+    raw_production_mix: pd.DataFrame, logger: Logger
+) -> List[dict]:
+    production_mix = []
+    generation_units = set(raw_production_mix.columns)
+    generation_units.remove("Period Start")
+    generation_units.remove("Period End")
+    if not generation_units == PLANT_MAPPING.keys():
+        raise Exception(
+            f"New generator {generation_units - PLANT_MAPPING.keys()} detected in AU-NT, please update the mapping of generators."
+        )
+    for _, production in raw_production_mix.iterrows():
+        data_point = {
+            "zoneKey": "AU-NT",
+            "datetime": production["Period Start"],
```

Comment:
This don't seem to be a native datetime object so the test_parser is throwing an exception on it. So we will have to convert it to a native datetime object first before we return it.